File: code/dart/test/plugin_host/plugin_host_guardrails_test.dart
import 'dart:convert';

import 'package:modular_api/modular_api.dart';
import 'package:modular_api/src/core/error_response_middleware.dart';
import 'package:modular_api/src/core/logger/logging_middleware.dart';
import 'package:modular_api/src/core/usecase/usecase_http_handler.dart';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import 'package:test/test.dart';

void main() {
  group('Plugin middleware guardrails', () {
    test('annotates the completed request log when plugin middleware short-circuits before the core handler', () async {
      final events = <String>[];
      final logOutput = StringBuffer();
      final handler = buildGuardrailHandler(
        plugins: [
          MiddlewarePlugin(
            id: 'acme.guard',
            definitions: [
              MiddlewareDefinition(
                id: 'auth',
                slot: 'preHandler',
                order: 0,
                handler: (nextHandler) {
                  return (request) async => Response(
                        401,
                        headers: {'content-type': 'application/json'},
                        body: jsonEncode({'error': 'blocked by plugin'}),
                      );
                },
              ),
            ],
          ),
        ],
        events: events,
        logOutput: logOutput,
      );

      final response = await handler(
        Request(
          'POST',
          Uri.parse('http://localhost/api/demo/pipeline'),
          headers: {'X-Request-ID': 'trace-dart-short-circuit'},
          body: jsonEncode({'name': 'Ada'}),
        ),
      );

      expect(response.statusCode, 401);
      expect(jsonDecode(await response.readAsString()), {'error': 'blocked by plugin'});
      expect(events, isEmpty);

      final lines = logOutput.toString().trim().split('\n');
      final completedLog = jsonDecode(lines.last) as Map<String, dynamic>;
      expect(completedLog['msg'], 'request completed');
      expect(completedLog['trace_id'], 'trace-dart-short-circuit');
      expect(completedLog['short_circuit'], isTrue);
      expect(completedLog['short_circuit_plugin_id'], 'acme.guard');
      expect(completedLog['short_circuit_middleware_id'], 'acme.guard.auth');
      expect(completedLog['short_circuit_slot'], 'preHandler');
    });

    test('returns a normalized 500 JSON response when plugin middleware throws outside the core handler', () async {
      final events = <String>[];
      final logOutput = StringBuffer();
      final handler = buildGuardrailHandler(
        plugins: [
          MiddlewarePlugin(
            id: 'acme.guard',
            definitions: [
              MiddlewareDefinition(
                id: 'boom',
                slot: 'preHandler',
                order: 0,
                handler: (nextHandler) {
                  return (request) async => throw StateError('boom');
                },
              ),
            ],
          ),
        ],
        events: events,
        logOutput: logOutput,
      );

      final response = await handler(
        Request(
          'POST',
          Uri.parse('http://localhost/api/demo/pipeline'),
          headers: {'X-Request-ID': 'trace-dart-error-guardrail'},
          body: jsonEncode({'name': 'Ada'}),
        ),
      );

      expect(response.statusCode, 500);
      expect(jsonDecode(await response.readAsString()), {'error': 'Internal server error'});
      expect(events, isEmpty);

      final lines = logOutput.toString().trim().split('\n');
      final parsedLogs = lines.map((line) => jsonDecode(line) as Map<String, dynamic>).toList();
      final errorLog = parsedLogs.firstWhere((line) => line['msg'] == 'Unhandled error in request pipeline');
      expect(errorLog['trace_id'], 'trace-dart-error-guardrail');
      expect(errorLog['level'], 'error');
      expect(errorLog['fields'], {'error': 'Bad state: boom', 'status': 500});

      final completedLog = parsedLogs.last;
      expect(completedLog['msg'], 'request completed');
      expect(completedLog['status'], 500);
      expect(completedLog.containsKey('short_circuit'), isFalse);
    });
  });
}

Handler buildGuardrailHandler({
  required List<Plugin> plugins,
  required List<String> events,
  required StringBuffer logOutput,
}) {
  final host = RuntimePluginHost(
    basePath: '/api',
    title: 'Guardrail Test API',
    version: '0.1.0',
  );

  for (final plugin in plugins) {
    host.beginPluginSetup(plugin.manifest.id);
    plugin.setup(host);
    host.endPluginSetup();
  }

  host.freeze();
  host.assertValid();

  final router = Router()
    ..post(
      '/api/demo/pipeline',
      useCaseHttpHandler(
        (json) => GuardrailUseCase(
          input: GuardrailInput.fromJson(json),
          events: events,
        ),
        inputExample: GuardrailInput(),
      ),
    );

  var pipeline = const Pipeline().addMiddleware(
    loggingMiddleware(
      logLevel: LogLevel.debug,
      serviceName: 'guardrail-test',
      sink: logOutput,
    ),
  );
  pipeline = pipeline.addMiddleware(errorResponseMiddleware());

  for (final middleware in host.middlewaresForSlot('preRouting')) {
    pipeline = pipeline.addMiddleware(middleware.handler);
  }
  for (final middleware in host.middlewaresForSlot('preHandler')) {
    pipeline = pipeline.addMiddleware(middleware.handler);
  }
  for (final middleware in host.middlewaresForSlot('postHandler')) {
    pipeline = pipeline.addMiddleware(middleware.handler);
  }

  return pipeline.addHandler(router.call);
}

class GuardrailInput extends Input {
  final String name;

  GuardrailInput({this.name = ''});

  factory GuardrailInput.fromJson(Map<String, dynamic> json) {
    return GuardrailInput(name: json['name']?.toString() ?? '');
  }

  @override
  Map<String, dynamic> toJson() => {'name': name};

  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'name': {'type': 'string'},
        },
        'required': ['name'],
      };
}

class GuardrailOutput extends Output {
  final String message;

  GuardrailOutput({this.message = 'ok'});

  @override
  int get statusCode => 200;

  @override
  Map<String, dynamic> toJson() => {'message': message};

  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'message': {'type': 'string'},
        },
        'required': ['message'],
      };
}

class GuardrailUseCase implements UseCase<GuardrailInput, GuardrailOutput> {
  @override
  final GuardrailInput input;

  final List<String> events;

  @override
  ModularLogger? logger;

  GuardrailUseCase({required this.input, required this.events});

  @override
  String? validate() {
    events.add('validate');
    return null;
  }

  @override
  Future<GuardrailOutput> execute() async {
    events.add('execute');
    return GuardrailOutput(message: 'Hello, ${input.name}');
  }
}

class MiddlewareDefinition {
  final String id;
  final String slot;
  final int order;
  final Middleware handler;

  const MiddlewareDefinition({
    required this.id,
    required this.slot,
    required this.order,
    required this.handler,
  });
}

class MiddlewarePlugin implements Plugin {
  @override
  final PluginManifest manifest;

  final List<MiddlewareDefinition> definitions;

  MiddlewarePlugin({
    required String id,
    required this.definitions,
  }) : manifest = PluginManifest(
          id: id,
          displayName: 'Middleware Plugin',
          version: '0.1.0',
          hostApiVersion: '>=0.1.0 <0.2.0',
        );

  @override
  void setup(PluginHost host) {
    for (final definition in definitions) {
      host.registerMiddleware(
        PluginMiddleware(
          id: '${manifest.id}.${definition.id}',
          slot: definition.slot,
          order: definition.order,
          handler: definition.handler,
        ),
      );
    }
  }
}

File: code/ts/modular_api/test/plugin_host/plugin_host.guardrails.test.ts
import express from 'express';
import request from 'supertest';
import { describe, expect, it } from 'vitest';

import { Input, Output, type Plugin, type PluginHost, type PluginManifest, type PluginMiddleware, type UseCase } from '../../src';
import { bodyParserErrorHandler } from '../../src/core/body_parser_error_handler';
import { LogLevel } from '../../src/core/logger/logger';
import { loggingMiddleware } from '../../src/core/logger/logging_middleware';
import { RuntimePluginHost } from '../../src/core/plugin';
import { useCaseHandler } from '../../src/core/usecase_handler';
import { unhandledRequestErrorHandler } from '../../src/core/unhandled_request_error_handler';

describe('Plugin middleware guardrails', () => {
  it('annotates the completed request log when plugin middleware short-circuits before the core handler', async () => {
    const events: string[] = [];
    const logLines: string[] = [];
    const app = buildGuardrailApp({
      logLines,
      events,
      plugins: [
        new MiddlewarePlugin('acme.guard', [
          {
            id: 'auth',
            slot: 'preHandler',
            order: 0,
            handler: (_req, res) => {
              res.status(401).json({ error: 'blocked by plugin' });
            },
          },
        ]),
      ],
    });

    const response = await request(app)
      .post('/api/demo/pipeline')
      .set('X-Request-ID', 'trace-ts-short-circuit')
      .send({ name: 'Ada' });

    expect(response.status).toBe(401);
    expect(response.body).toEqual({ error: 'blocked by plugin' });
    expect(events).toEqual([]);

    const completedLog = JSON.parse(logLines[logLines.length - 1]) as Record<string, unknown>;
    expect(completedLog.msg).toBe('request completed');
    expect(completedLog.trace_id).toBe('trace-ts-short-circuit');
    expect(completedLog.short_circuit).toBe(true);
    expect(completedLog.short_circuit_plugin_id).toBe('acme.guard');
    expect(completedLog.short_circuit_middleware_id).toBe('acme.guard.auth');
    expect(completedLog.short_circuit_slot).toBe('preHandler');
  });

  it('returns a normalized 500 JSON response when plugin middleware throws outside the core handler', async () => {
    const events: string[] = [];
    const logLines: string[] = [];
    const app = buildGuardrailApp({
      logLines,
      events,
      plugins: [
        new MiddlewarePlugin('acme.guard', [
          {
            id: 'boom',
            slot: 'preHandler',
            order: 0,
            handler: () => {
              throw new Error('boom');
            },
          },
        ]),
      ],
    });

    const response = await request(app)
      .post('/api/demo/pipeline')
      .set('X-Request-ID', 'trace-ts-error-guardrail')
      .send({ name: 'Ada' });

    expect(response.status).toBe(500);
    expect(response.body).toEqual({ error: 'Internal server error' });
    expect(events).toEqual([]);

    const parsedLogs = logLines.map((line) => JSON.parse(line) as Record<string, unknown>);
    const errorLog = parsedLogs.find((line) => line.msg === 'Unhandled error in request pipeline');
    expect(errorLog).toMatchObject({
      trace_id: 'trace-ts-error-guardrail',
      level: 'error',
      fields: { status: 500, error: 'Error: boom' },
    });

    const completedLog = parsedLogs[parsedLogs.length - 1];
    expect(completedLog.msg).toBe('request completed');
    expect(completedLog.status).toBe(500);
    expect(completedLog.short_circuit).toBeUndefined();
  });
});

function buildGuardrailApp(options: {
  plugins: Plugin[];
  events: string[];
  logLines: string[];
}) {
  const app = express();
  const host = new RuntimePluginHost({
    basePath: '/api',
    title: 'Guardrail Test API',
    version: '0.1.0',
  });

  for (const plugin of options.plugins) {
    host.beginPluginSetup(plugin.manifest.id);
    plugin.setup(host);
    host.endPluginSetup();
  }

  host.freeze();
  host.assertValid();

  app.use(
    loggingMiddleware({
      logLevel: LogLevel.debug,
      serviceName: 'guardrail-test',
      writeFn: (line) => options.logLines.push(line),
    }),
  );
  app.use(express.json());
  app.use(bodyParserErrorHandler);

  host.applyMiddlewares('preRouting', app);
  host.applyMiddlewares('preHandler', app);
  host.applyMiddlewares('postHandler', app);

  app.post(
    '/api/demo/pipeline',
    useCaseHandler((json) => new GuardrailUseCase(GuardrailInput.fromJson(json), options.events), {
      inputClass: GuardrailInput,
    }),
  );

  app.use(unhandledRequestErrorHandler);
  return app;
}

class GuardrailInput extends Input {
  constructor(public name = '') {
    super();
  }

  static fromJson(json: Record<string, unknown>): GuardrailInput {
    return new GuardrailInput(typeof json.name === 'string' ? json.name : '');
  }

  override toJson(): Record<string, unknown> {
    return { name: this.name };
  }

  override toSchema(): Record<string, unknown> {
    return {
      type: 'object',
      properties: {
        name: { type: 'string' },
      },
      required: ['name'],
    };
  }
}

class GuardrailOutput extends Output {
  constructor(public message = 'ok') {
    super();
  }

  override get statusCode(): number {
    return 200;
  }

  override toJson(): Record<string, unknown> {
    return { message: this.message };
  }

  override toSchema(): Record<string, unknown> {
    return {
      type: 'object',
      properties: {
        message: { type: 'string' },
      },
      required: ['message'],
    };
  }
}

class GuardrailUseCase implements UseCase<GuardrailInput, GuardrailOutput> {
  logger = undefined;

  constructor(
    readonly input: GuardrailInput,
    private readonly events: string[],
  ) {}

  validate(): string | null {
    this.events.push('validate');
    return null;
  }

  async execute(): Promise<GuardrailOutput> {
    this.events.push('execute');
    return new GuardrailOutput(`Hello, ${this.input.name}`);
  }
}

class MiddlewarePlugin implements Plugin {
  readonly manifest: PluginManifest;

  constructor(
    id: string,
    private readonly definitions: Array<{
      id: string;
      slot: PluginMiddleware['slot'];
      order: number;
      handler: PluginMiddleware['handler'];
    }>,
  ) {
    this.manifest = {
      id,
      displayName: 'Middleware Plugin',
      version: '0.1.0',
      hostApiVersion: '>=0.1.0 <0.2.0',
    };
  }

  setup(host: PluginHost): void {
    for (const definition of this.definitions) {
      host.registerMiddleware({
        id: `${this.manifest.id}.${definition.id}`,
        slot: definition.slot,
        order: definition.order,
        handler: definition.handler,
      });
    }
  }
}